src/calcium/ca.jl

```diff
@@ -0,0 +1,1447 @@
+extension numbers; otherwise, the rewriting is only applied to the
+top-level extension numbers.
+
+The result is not a normal form in the strong sense (the same number
+can have many possible representations even after applying this
+transformation), but this transformation can nevertheless be a useful
+heuristic for simplification.
+"""
+function complex_normal_form(a::ca; deep::Bool=true)
+   C = a.parent
+   r = C()
+   ccall((:ca_rewrite_complex_normal_form, libcalcium), Nothing,
+         (Ref{ca}, Ref{ca}, Cint, Ref{CalciumField}), r, a, deep, C)
+   check_special(r)
+   return r
+end
+
+###############################################################################
+#
+#   Conversions
+#
+###############################################################################
+
+function fmpq(a::ca)
+   C = a.parent
+   res = fmpq()
+   ok = Bool(ccall((:ca_get_fmpq, libcalcium), Cint,
+        (Ref{fmpq}, Ref{ca}, Ref{CalciumField}), res, a, C))
+   !ok && error("unable to convert to a rational number")
+   return res
+end
+
+function fmpz(a::ca)
+   C = a.parent
+   res = fmpz()
+   ok = Bool(ccall((:ca_get_fmpz, libcalcium), Cint,
+        (Ref{fmpz}, Ref{ca}, Ref{CalciumField}), res, a, C))
+   !ok && error("unable to convert to an integer")
+   return res
+end
+
+function qqbar(a::ca)
+   C = a.parent
+   res = qqbar()
+   ok = Bool(ccall((:ca_get_qqbar, libcalcium), Cint,
+        (Ref{qqbar}, Ref{ca}, Ref{CalciumField}), res, a, C))
+   !ok && error("unable to convert to an algebraic number")
+   return res
+end
+
+(R::FlintRationalField)(a::ca) = fmpq(a)
+(R::FlintIntegerRing)(a::ca) = fmpz(a)
+(R::CalciumQQBarField)(a::ca) = qqbar(a)
+
+function (R::AcbField)(a::ca; parts::Bool=false)
+   C = a.parent
+   prec = precision(R)
+   z = R()
+   if parts
+      ccall((:ca_get_acb_accurate_parts, libcalcium),
+        Nothing, (Ref{acb}, Ref{ca}, Int, Ref{CalciumField}), z, a, prec, C)
+   else
+      ccall((:ca_get_acb, libcalcium),
+        Nothing, (Ref{acb}, Ref{ca}, Int, Ref{CalciumField}), z, a, prec, C)
+   end
+   return z
+end
+
+function (R::ArbField)(a::ca; check::Bool=true)
+   C = a.parent
+   prec = precision(R)
+   if check
+      z = AcbField(prec)(a, parts=true)
+      if isreal(z)
+         return real(z)
+      else
+         error("unable to convert to a real number")
+      end
+   else
+      z = AcbField(prec)(a, parts=false)
+      if accuracy_bits(real(z)) < prec - 5
+          z = AcbField(prec)(a, parts=true)
+      end
+      return real(z)
+   end
+end
+
+###############################################################################
+#
+#   Unsafe functions
+#
+###############################################################################
+
+function zero!(z::ca)
+   C = z.parent
+   ccall((:ca_zero, libcalcium), Nothing, (Ref{ca}, Ref{CalciumField}), z, C)
+   return z
+end
+
+function mul!(z::ca, x::ca, y::ca)
+   if z.parent != x.parent || x.parent != y.parent
```

Comment:
We do not check parents in unsafe operators.

Reply:
Nonetheless, this check did help catch a bug. I'd rather keep it until the code is more battle-tested.